addons/account_asset/account_asset.py

```diff
@@ -1,351 +1,325 @@
+        help="Choose the method to use to compute the dates and number of depreciation lines.\n"
+           "  * Number of Depreciations: Fix the number of depreciation lines and the time between 2 depreciations.\n" \
+           "  * Ending Date: Choose the time between 2 depreciations and the date the depreciations won't go beyond.")
+    method_end = fields.Date('Ending date')
+    prorata = fields.Boolean(string='Prorata Temporis', help='Indicates that the first depreciation entry for this asset have to be done from the purchase date instead of the first January')
+    open_asset = fields.Boolean(string='Skip Draft State', help="Check this if you want to automatically confirm the assets of this category when created by invoices.")
+    type = fields.Selection([('sale', 'Sale: Revenue Recognition'), ('purchase', 'Purchase: Asset')], required=True, index=True, default='purchase')
+
+    @api.onchange('account_asset_id')
+    def onchange_account_asset(self):
+        self.account_depreciation_id = self.account_asset_id
+
+    @api.onchange('type')
+    def onchange_type(self):
+        if self.type == 'sale':
+            self.prorata = True
+            self.method_period = 1
+        else:
+            self.method_period = 12
 
 
-class account_asset_asset(osv.osv):
+class account_asset_asset(models.Model):
     _name = 'account.asset.asset'
-    _description = 'Asset'
-
-    def unlink(self, cr, uid, ids, context=None):
-        for asset in self.browse(cr, uid, ids, context=context):
-            if asset.account_move_line_ids: 
-                raise UserError(_('You cannot delete an asset that contains posted depreciation lines.'))
-        return super(account_asset_asset, self).unlink(cr, uid, ids, context=context)
-
-    def _get_period(self, cr, uid, context=None):
-        periods = self.pool.get('account.period').find(cr, uid, context=context)
+    _description = 'Asset/Revenue Recognition'
+    _inherit = ['mail.thread', 'ir.needaction_mixin']
+
+    account_move_line_ids = fields.One2many('account.move.line', 'asset_id', string='Entries', readonly=True, states={'draft': [('readonly', False)]})
+    entry_count = fields.Integer(compute='_entry_count', string='# Asset Entries')
+    name = fields.Char(string='Asset/Deferred Revenue Name', required=True, readonly=True, states={'draft': [('readonly', False)]})
+    code = fields.Char(string='Reference', size=32, readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env['ir.sequence'].next_by_code('account.asset.code'))
+    value = fields.Float(string='Gross Value', required=True, readonly=True, digits=dp.get_precision('Account'), states={'draft': [('readonly', False)]})
+    currency_id = fields.Many2one('res.currency', string='Currency', required=True, readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env['res.users'].browse(self._uid).company_id.currency_id.id)
```

Comment:
Use:
default=lambda self: self.env.user.company_id.currency_id.id
Instead of:
default=lambda self: self.env['res.users'].browse(self._uid).company_id.currency_id.id